File: openmole/core/org.openmole.core.implementation/src/main/java/org/openmole/core/implementation/execution/batch/BatchServiceGroup.java
/*
 *  Copyright (C) 2010 reuillon
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.core.implementation.execution.batch;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.Semaphore;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

import org.openmole.commons.exception.InternalProcessingError;
import org.openmole.core.implementation.internal.Activator;
import org.openmole.core.model.execution.batch.IAccessToken;
import org.openmole.core.model.execution.batch.IBatchService;
import org.openmole.core.model.execution.batch.IBatchServiceGroup;
import org.openmole.commons.aspect.eventdispatcher.IObjectChangedSynchronousListener;
import org.openmole.commons.exception.UserBadDataError;
import org.openmole.commons.tools.service.Priority;
import org.openmole.commons.tools.service.RNG;
import org.openmole.core.batchservicecontrol.IUsageControl;
import scala.Tuple2;

public class BatchServiceGroup<T extends IBatchService> implements IBatchServiceGroup<T> {

    class BatchRessourceGroupAdapter implements IObjectChangedSynchronousListener<IUsageControl> {

        @Override
        public void objectChanged(IUsageControl obj) throws InternalProcessingError, UserBadDataError {
            ressourceTokenReleased();
        }
        
    }

    final ArrayList<T> resources = new ArrayList<T>();

    transient Semaphore waiting;
    transient Lock selectingRessource;
    
    final private int expulseThreshold;

    public BatchServiceGroup(int threshold) {
        super();
        this.expulseThreshold = threshold;
    }

    private synchronized Semaphore getWaiting() {
        if (waiting == null) {
            waiting = new Semaphore(0);
        }
        return waiting;
    }

    void waitForRessourceReleased() throws InterruptedException {
        getWaiting().acquire();
    }

    @Override
    public Tuple2<T, IAccessToken> getAService() throws InterruptedException, InternalProcessingError, UserBadDataError {

        getSelectingRessource().lock();
        try {
            Tuple2<T, IAccessToken> ret = null;

            while (ret == null) {
                ArrayList<T> resourcesCopy; 

                //Select the less failing resources
                synchronized (resources) {

                    Iterator<T> resourcesIt = resources.iterator();

                    while (resourcesIt.hasNext()) {
                        IBatchService resource = resourcesIt.next();

                        if (Activator.getBatchRessourceControl().getController(resource.getDescription()).getFailureControl().getFailureRate() > getExpulseThreshold()) {
                            resourcesIt.remove();
                        }
                    }

                    if(resources.isEmpty()) throw new InternalProcessingError("No more reliable resource available.");
                    
                    resourcesCopy = new ArrayList<T>(resources.size());
                    resourcesCopy.addAll(resources);
                }

                //Among them select one not over loaded
                Iterator<T> bestResourcesIt = resourcesCopy.iterator();
                List<Tuple2<T, IAccessToken>> notLoaded = new ArrayList<Tuple2<T, IAccessToken>>();

                while (bestResourcesIt.hasNext()) {
                    
                    T cur = bestResourcesIt.next();

                    IAccessToken token = Activator.getBatchRessourceControl().getController(cur.getDescription()).getUsageControl().getAccessTokenInterruptly();

                    if (token != null) {
                        notLoaded.add(new Tuple2<T, IAccessToken>(cur, token));
                    }
                }
               
                if (notLoaded.size() > 0) {
                    ret = notLoaded.remove(RNG.getRng().nextInt(notLoaded.size()));

                    for (Tuple2<T, IAccessToken> other : notLoaded) {                    
                         Activator.getBatchRessourceControl().getController(other._1().getDescription()).getUsageControl().releaseToken(other._2());     
                    }
                } else {
                    waitForRessourceReleased();
                }

            }
           return ret;
        } finally {
            getSelectingRessource().unlock();
        }
    }

    @Override
    public void add(T service) {
       synchronized (resources) {
            resources.add(service);
            IUsageControl usageControl = Activator.getBatchRessourceControl().getController(service.getDescription()).getUsageControl();
            Activator.getEventDispatcher().registerListener(usageControl, Priority.NORMAL.getValue(), new BatchRessourceGroupAdapter(), IUsageControl.resourceReleased);
        }
        ressourceTokenReleased();
    }
    
    @Override
    public void addAll(Iterable<? extends T> services) {
        synchronized (resources) {
            for(T service: services) {
                add(service);
            }
        }
    }
    
    

    public T get(int index) {
        synchronized (resources) {
            return resources.get(index);
        }
    }

    public boolean isEmpty() {
        return resources.isEmpty();
    }

    @Override
    public int size() {
        return resources.size();
    }

    @Override
    public Iterator<T> iterator() {
        return resources.iterator();
    }

    private int getExpulseThreshold() {
        return expulseThreshold;
    }

    public Lock getSelectingRessource() {
        if (selectingRessource != null) {
            return selectingRessource;
        }
        synchronized (this) {
            if (selectingRessource == null) {
                selectingRessource = new ReentrantLock();
            }
        }
        return selectingRessource;
    }


    private void ressourceTokenReleased() {
        getWaiting().release();
    }
}

File: openmole/core/org.openmole.core.implementation/src/main/scala/org/openmole/core/implementation/execution/batch/BatchServiceGroup.scala
/*
  * Copyright (C) 2010 reuillon
  *
  * This program is free software: you can redistribute it and/or modify
  * it under the terms of the GNU Affero General Public License as published by
  * the Free Software Foundation, either version 3 of the License, or
  * (at your option) any later version.
  *
  * This program is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
  *
  * You should have received a copy of the GNU General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */

package org.openmole.core.implementation.execution.batch

import java.util.ArrayList
import java.util.LinkedList
import java.util.concurrent.Semaphore
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock

import org.openmole.commons.aspect.eventdispatcher.IObjectListener
import org.openmole.commons.exception.InternalProcessingError
import org.openmole.commons.tools.service.RNG
import org.openmole.commons.tools.service.Priority

import org.openmole.core.batchservicecontrol.IQualityControl
import org.openmole.core.batchservicecontrol.IUsageControl
import org.openmole.core.model.execution.batch.IBatchServiceGroup
import org.openmole.core.implementation.internal.Activator
import org.openmole.core.model.execution.batch.IAccessToken
import org.openmole.core.model.execution.batch.IBatchService
import scala.collection.mutable.ArrayBuffer
import scala.collection.mutable.ListBuffer
import scala.collection.JavaConversions._

class BatchServiceGroup[T <: IBatchService[_,_]](expulseThreshold: Int) extends IBatchServiceGroup[T] {

  class BatchRessourceGroupAdapterUsage extends IObjectListener[IUsageControl] {
    override def eventOccured(obj: IUsageControl) = waiting.release
  }
  
  private val resources = new LinkedList[T]

  @transient lazy val waiting: Semaphore = new Semaphore(0)
  @transient lazy val selectingRessource: Lock = new ReentrantLock

  override def selectAService: (T, IAccessToken) = {
    selectingRessource.lock
    try {
      var ret: (T, IAccessToken) = null

      while (ret == null) {
        //Select the less failing resources
        val resourcesCopy = resources.synchronized {
          val resourcesIt = resources.iterator

          while (resourcesIt.hasNext) {
            val resource = resourcesIt.next

            Activator.getBatchRessourceControl.qualityControl(resource.description) match {
              case Some(f) => if(f.failureRate > expulseThreshold) resourcesIt.remove
              case None => 
            }
          }

          if(resources.isEmpty) throw new InternalProcessingError("No more reliable resource available.");
                    
          val ret = new ArrayList[T](resources.size)
          ret.addAll(resources)
          ret
        }

        //Among them select one not over loaded
        val bestResourcesIt = resourcesCopy.iterator
        val notLoaded = new ArrayBuffer[(T, IAccessToken, Long)]
        var totalQuality = 0L
        
        while (bestResourcesIt.hasNext) {       
          val cur = bestResourcesIt.next

            Activator.getBatchRessourceControl.usageControl(cur.description).tryGetToken match {
              case None =>
              case Some(token) =>
                val toInsert = ((cur, token, 
                 Activator.getBatchRessourceControl.qualityControl(cur.description) match {
                  case None => 1L
                  case Some(quality) => quality.quality
                }))
                notLoaded += toInsert
                totalQuality += toInsert._3
            }
        }
               
        if (notLoaded.size > 0) {
          var selectedIndex = RNG.nextLong(totalQuality)
          val it = notLoaded.iterator
          
          var selected = it.next
          
          while(selectedIndex >= selected._3) {
            selectedIndex -= selected._3
            selected = it.next
          }
          
          for (service <- notLoaded) {    
            if(service._1.description != selected._1.description) Activator.getBatchRessourceControl.usageControl(service._1.description).releaseToken(service._2) 
          }
          ret = (selected._1, selected._2)
        } else {
          waiting.acquire
        }
      }
      return ret
    } finally {
      selectingRessource.unlock
    }
  }

  override def add(service: T) = {
    resources.synchronized {
      resources.add(service)
      val usageControl = Activator.getBatchRessourceControl.usageControl(service.description)
      Activator.getEventDispatcher.registerForObjectChangedSynchronous(usageControl, Priority.NORMAL, new BatchRessourceGroupAdapterUsage, IUsageControl.ResourceReleased)
    }
    waiting.release
  }
    
  override def addAll(services: Iterable[T]) {
    resources.synchronized {
      for(service <- services) {
        add(service)
      }
    }
  }

  def get(index: Int): T = {
    resources.synchronized {
      resources.get(index);
    }
  }

  override def isEmpty: Boolean = resources.isEmpty

  override def size: Int = resources.size

  override def iterator: Iterator[T] = resources.iterator

}
